state/protocol/inmem/epoch.go

```diff
@@ -12,121 +12,34 @@ import (
-func (e Epoch) DKGPhase1FinalView() (uint64, error) { return e.enc.DKGPhase1FinalView, nil }
-func (e Epoch) DKGPhase2FinalView() (uint64, error) { return e.enc.DKGPhase2FinalView, nil }
-func (e Epoch) DKGPhase3FinalView() (uint64, error) { return e.enc.DKGPhase3FinalView, nil }
-func (e Epoch) FinalView() (uint64, error)          { return e.enc.FinalView, nil }
-func (e Epoch) InitialIdentities() (flow.IdentitySkeletonList, error) {
-	return e.enc.InitialIdentities, nil
-}
-func (e Epoch) RandomSource() ([]byte, error) {
-	return e.enc.RandomSource, nil
-}
-
-// TargetDuration returns the desired real-world duration for this epoch, in seconds.
-// This target is specified by the FlowEpoch smart contract in the EpochSetup event
-// and used by the Cruise Control system to moderate the block rate.
-func (e Epoch) TargetDuration() (uint64, error) {
-	return e.enc.TargetDuration, nil
-}
-
-// TargetEndTime returns the desired real-world end time for this epoch, represented as
-// Unix Time (in units of seconds). This target is specified by the FlowEpoch smart contract in
-// the EpochSetup event and used by the Cruise Control system to moderate the block rate.
-func (e Epoch) TargetEndTime() (uint64, error) {
-	return e.enc.TargetEndTime, nil
-}
-
-func (e Epoch) Clustering() (flow.ClusterList, error) {
-	return e.enc.Clustering, nil
-}
-
-func (e Epoch) DKG() (protocol.DKG, error) {
-	if e.enc.DKG != nil {
-		return DKG{*e.enc.DKG}, nil
-	}
-	return nil, protocol.ErrNextEpochNotCommitted
-}
-
-func (e Epoch) Cluster(i uint) (protocol.Cluster, error) {
-	if e.enc.Clusters == nil {
-		return nil, protocol.ErrNextEpochNotCommitted
-	}
-
-	if i >= uint(len(e.enc.Clusters)) {
-		return nil, fmt.Errorf("no cluster with index %d: %w", i, protocol.ErrClusterNotFound)
-	}
-	return Cluster{e.enc.Clusters[i]}, nil
-}
-
-func (e Epoch) ClusterByChainID(chainID flow.ChainID) (protocol.Cluster, error) {
-	if e.enc.Clusters == nil {
-		return nil, protocol.ErrNextEpochNotCommitted
-	}
-
-	for _, cluster := range e.enc.Clusters {
-		if cluster.RootBlock.Header.ChainID == chainID {
-			return Cluster{cluster}, nil
-		}
-	}
-	chainIDs := make([]string, 0, len(e.enc.Clusters))
-	for _, cluster := range e.enc.Clusters {
-		chainIDs = append(chainIDs, string(cluster.RootBlock.Header.ChainID))
-	}
-	return nil, fmt.Errorf("no cluster with the given chain ID %v, available chainIDs %v: %w", chainID, chainIDs, protocol.ErrClusterNotFound)
-}
-
-func (e Epoch) FinalHeight() (uint64, error) {
-	if e.enc.FinalHeight != nil {
-		return *e.enc.FinalHeight, nil
-	}
-	return 0, protocol.ErrEpochTransitionNotFinalized
-}
-
-func (e Epoch) FirstHeight() (uint64, error) {
-	if e.enc.FirstHeight != nil {
-		return *e.enc.FirstHeight, nil
-	}
-	return 0, protocol.ErrEpochTransitionNotFinalized
-}
-
+// Epochs provides access to epoch data, backed by a rich protocol state entry.
 type Epochs struct {
-	enc EncodableEpochs
+	entry flow.RichProtocolStateEntry
```

Comment:
```suggestion
	*flow.RichProtocolStateEntry
```
any specific reason for using value? Data locality is not a thing as well as copy guarantees since `RichProtocolStateEntry` has many pointers. How about just embedding the pointer?

Reply:
Regarding using value over pointer, see [Alex's comment](https://github.com/onflow/flow-go/pull/5682#discussion_r1581713947) (it was a pointer originally, but I agree with Alex's point). 

Regarding embedding, in general I'd rather use a private field than a public embedded field by default, unless there is a specific reason to make everything public.